### PR TITLE
[WIP] Added support for BERT.

### PR DESCRIPTION
**Pull request by eaplatanios**
This PR adds support for BERT. **It is not ready to merge yet.**

I believe that the following features should be moved to `swift-apis` if people agree:
- A new `Embedding` that ought to support using matrix multiplications instead of gather ops for faster execution on TPUs (this is currently commented out due an AutoDiff bug). This can replace the existing layer in swift-apis and should be backwards compatible.
- A `MultiHeadAttention` layer that can be used for other models too (other than BERT or Transformers).
- A `Regularizable` protocol that I don't really like, but it's a temporary solution for supporting weight decay in BERT. We probably shouldn't move this to swift-apis until we have thought out a nicer solution.
- A new `Optimizer` protocol that supports learning rate schedules and a  `WeightDecayedAdam` implementation.
- `ScheduledParameter` corresponds to tensorflow/swift-apis#431.

The following features are also added:
- Multiple text tokenization approaches, including a byte-pair encoding (BPE) tokenizer as well as a WordPiece tokenizer.
- A Transformer implementation.
- A BERT implementation that supports multiple variants (i.e., BERT, RoBERTa, and ALBERT).
- Support for automatically downloading and loading pre-trained models for all these variants.

Note that for RoBERTa I had to convert the published PyTorch checkpoints to TF checkpoints compatible with this model. I have uploaded these to my Dropbox account, but they take too much space, so I would really appreciate it if we could move them to Google Cloud Storage.

This is not ready to merge yet, but is rather aimed at getting feedback to improve the API and also making sure that the code style is compatible with this repository. A big open question is, how do we want to test this? I know it's working fine as I'm using it for my own research projects, but I don't know what kind of tests would be appropriate.

Sorry if I missed something in this list of changes. I'll keep updating it as we refine the PR.

**NOTE:** The new layers are not generic over the `Scalar` type due to a compiler bug ([TF-427](https://bugs.swift.org/browse/TF-427)).

**NOTE:** Compilation is currently failing because the synthesized `TangentVector` initializers are internal and I cannot declare conformances to `Regularizable` without using those initializers.

cc @saeta @dan-zheng @BradLarson

### CODE REVIEW

**Comment by Shashi456** [comment]
This is absolutely huge @eaplatanios. I've been working on BERT for the past few months and can't wait to check this code out :).

File: Models/Text/Initializers.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        Tensor<Scalar>(
+            randomTruncatedNormal: $0,
+            mean: mean,
+            standardDeviation: standardDeviation,
+            seed: seed)
+    }
+}
+
+extension Tensor where Scalar: TensorFlowFloatingPoint {
```
Comment (BradLarson) [inline Models/Text/Initializers.swift]: I know this is a WIP, but doesn't the truncated normal initializer already exist within swift-apis: https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Initializers.swift#L444 ?

Comment (eaplatanios) [reply Models/Text/Initializers.swift]: Oops you're right. I forgot this got merged earlier :) I removed it from this PR.

**Comment by Shashi456** [comment]
@eaplatanios What do you think about restructuring the encoding techniques under a `preproccessing` directory either in this repo or the `swift-apis` repo, These techniques are used everywhere. 
I'm also suggesting for a preprocessing directory, because down the line if we image preprocessing techniques including augmentatation, this structure would benefit that as well.
cc: @BradLarson @saeta

**Comment by Shashi456** [comment]
I think we could take some inspiration on how to test these models from how these are done in pytorch-transformers, as @saeta mentioned in the design meeting, they do just check loading the model and check if the tensor shapes are right. You can take a look at these tests [here](https://github.com/huggingface/transformers/blob/master/transformers/tests/modeling_bert_test.py).

**Comment by BradLarson** [comment]
I'd like to see what can be done to help drive this to completion. I forget where we left things after the community meeting, but is the biggest blocker right now the compilation errors around TangentVector and the activation functions in Attention.swift?

If there are areas you'd like me to look into, I'd be glad to do so.

**Comment by eaplatanios** [comment]
Thanks @BradLarson ! The only blocker is the `TangentVector` initializer. The issues related to attention have to do with supporting generic BERT layers (currently they're just `Float`-valued).

**Comment by eaplatanios** [comment]
Please also let me know what you guys think should be pushed to swift-apis and what kind of restructuring and testing would be useful. :)

File: Models/Text/WeightDecayedAdam.swift [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+public protocol Regularizable: Differentiable {
+    /// The contribution of this term to the regularization term. This should be set to
+    /// `TangentVector.zero` if this term should not contribute to the regularization term
+    /// (e.g., for layer normalization parameters).
+    var regularizationValue: TangentVector { get }
+}
+
+extension Dense: Regularizable {
+    public var regularizationValue: TangentVector {
+        // TODO: This initializer is currently internal.
```
Comment (dan-zheng) [inline Models/Text/WeightDecayedAdam.swift]: I filed [TF-1077](https://bugs.swift.org/browse/TF-1077) to track the non-public `TangentVector` memberwise initializer issue.

Comment (eaplatanios) [reply Models/Text/WeightDecayedAdam.swift]: That's great! Thanks a lot Dan! Also, what do you think of this solution for regularization? I don't really like it too much, but I also couldn't think of another easy way to support it. :/

File: Models/Text/BERT.swift [hunk 3]
```diff
@@ -0,0 +1,902 @@
...
+            case .albertLarge: return "ALBERT Large"
+            case .albertXLarge: return "ALBERT xLarge"
+            case .albertXXLarge: return "ALBERT xxLarge"
+            }
+        }
+
+        /// The URL where this pre-trained model can be downloaded from.
+        public var url: URL {
+            let bertPrefix = "https://storage.googleapis.com/bert_models/2018_"
+            let robertaPrefix = "https://www.dropbox.com/s"
```
Comment (BradLarson) [inline Models/Text/BERT.swift]: So that you don't have to store these large weights in Dropbox, I've re-uploaded them to a hosted GCS bucket we've created for weights / datasets:

https://storage.googleapis.com/s4tf-hosted-binaries/checkpoints/Text/RoBERTa/base.zip
https://storage.googleapis.com/s4tf-hosted-binaries/checkpoints/Text/RoBERTa/large.zip

If there are others you'd like me to place there, just let me know and I'll add them.

Comment (Shashi456) [reply Models/Text/BERT.swift]: @BradLarson would this bucket also be useful for adding support for pre-trained models and their weights?

Comment (BradLarson) [reply Models/Text/BERT.swift]: @Shashi456 - Yes, that's one of the goals, in addition to being a reliable source for sometimes-flaky dataset download locations. I'm working on a quick addition to the checkpoint loaders to make it easy to download from here, so that we simplify the process of working with models that need pretrained checkpoints (the existing Transformer and MiniGo models, as well as BERT and family) and start CI testing inference accuracy using real pretrained models.

**Comment by BradLarson** [comment]
Now that the TangentVector visibility issues have been resolved, is there anything else blocking the core model? If so, I'd love to see what we could do to resolve that.

Beyond the core model, do you have an example of BERT in action or a unit test of it that we can use to verify correct operation? If those are too tied up in any internal infrastructure you have, we could potentially pull this in and I could add a demo and / or tests as a follow-on, but if you have simple demo code for this that would really ease the process of pulling this in.

We can merge together the existing Transformer demo and the new Transformer model and utility functions you have here to migrate our text generation demo over to this new structure, but didn't know if there was a similar demo you had available for BERT.

**Comment by dan-zheng** [comment]
I merged the `tensorflow:bert-wip` branch into `eaplatanios:bert`. This includes the following changes:
- Some merge conflicts were fixed.
- Some differentiation-related compilation errors were fixed.
- A training loop for BERT was added in `Models/Text/main.swift`.
  - A [`BERT.PreTrainedModel.bertBase`](https://github.com/tensorflow/swift-models/blob/b7bcaa46bbf8de8fff04e03a0d4b47fa55c1b4f5/Models/Text/main.swift#L4) pretrained model is loaded and trains on the [CoLA task](https://github.com/tensorflow/swift-models/blob/bert-wip/Models/Text/CoLA.swift). You can run it via `swift run TextModels`.

There are a bunch of todos:
- Rewrite utilities for downloading/extracting data in this PR using unified `ModelSupport` APIs.
  - Consider generalizing [`DatasetUtilities.downloadResource`](https://github.com/tensorflow/swift-models/blob/b7bcaa46bbf8de8fff04e03a0d4b47fa55c1b4f5/Models/Text/CoLA.swift#L86-L93) to support URLs that don't end in the zipped file name. The [CoLA dataset URL](https://github.com/tensorflow/swift-models/blob/b7bcaa46bbf8de8fff04e03a0d4b47fa55c1b4f5/Models/Text/CoLA.swift#L182-L184) is one such example. Alternatively, we could host the dataset elsewhere.
- Improve code organization.
  - Currently, most BERT code lives at the top-level in `Models/Text`. `Models/Text` could be better organized, like `Models/ImageClassification`.
- Verify that BERT training converges, and that results match a reference implementation.
- Verify that other BERT variants work.
  - BERT variants like [`roberta`](https://github.com/tensorflow/swift-models/blob/b7bcaa46bbf8de8fff04e03a0d4b47fa55c1b4f5/Models/Text/BERT.swift#L367-L369) and [`albert`](https://github.com/tensorflow/swift-models/blob/b7bcaa46bbf8de8fff04e03a0d4b47fa55c1b4f5/Models/Text/BERT.swift#L371-L373) were added in this PR but are untested.

---

`swift run TextModels` works with latest Swift for TensorFlow toolchains:
```console
$ swift run TextModels
...
[Step: 41]	Loss: 0.3939771
[Step: 42]	Loss: 0.7205323
[Step: 43]	Loss: 0.4848549
[Step: 44]	Loss: 0.2727359
[Step: 45]	Loss: 0.3460037
[Step: 46]	Loss: 0.9574833
[Step: 47]	Loss: 0.7397041
[Step: 48]	Loss: 0.35861257
[Step: 49]	Loss: 0.4990799
[Step: 50]	Loss: 0.92075384
Evaluate BERT for the CoLA task:
Total predictions: 1043
True positives: 719
True negatives: 34
False positives: 288
False negatives: 34
["matthewsCorrelationCoefficient": 0.26019016]
```

If there are no objections, now seems like a good point to merge this PR and continue incremental improvements. Thanks @eaplatanios for driving this work!

**Comment by eaplatanios** [comment]
Thanks @dan-zheng and @BradLarson for getting this in! I'm sorry I haven't had too much time lately to address some of the mentioned todos.
